### PR TITLE
adds snapshot capabilities to logical replication iterator

### LINKED ISSUE #15: Snapshot support in logrepl mode
When using logical replication mode it should be possible to create a snapshot first.

### PR DESCRIPTION
**Pull request by dylanlott**
### Description

This PR refactors the `CDCIterator` into a combined `Iterator` that takes
a snapshot if configured to do so and then switches into logical replication
mode at its anchored LSN position.

Fixes #15

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-postgres/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

**Review by lovromazgon — CHANGES_REQUESTED** [review]
I want to point out two things.

First and foremost, there is a race condition. With this approach, we can't guarantee we won't miss records created between beginning the snapshot transaction and fetching the current LSN. It's a narrow window, nevertheless, it is a race and on a busy database we will hit it. After checking Debezium, it seems like they create the replication slot before starting the snapshot (they even have a [comment](https://github.com/debezium/debezium/blob/2d14b5b09f4af68f6070fbfc4e4a62cd323290f6/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java#L159-L162) related to this) and use the snapshot to initialize the snapshot iterator.

Now, the question is how do we handle the case when the replication slot already exists and we don't get a snapshot when creating the subscription? If I'm not mistaken, in that case Debezium falls back to the behavior you have implemented right now - they fetch the current LSN, do the snapshot and start processing logrepl messages from that LSN forward. As far as I understand it's a fallback that _mostly_ works. We shouldn't do that by default though, when there is a better way.

The second thing is that `Iterator` now combines the logic of the old CDC iterator and the combined iterator. IMO we should strive to keep those separate (e.g. have a `SnapshotIterator` a `CDCIterator` and an `Iterator` which ties them together), it will make it easier to reason about the logic and test it in separation. If we separate them, we will have a nice place where we can put the logic about choosing the correct starting LSN/snapshot (i.e. using the replication slot snapshot if present and falling back to the current LSN if it's not).

EDIT: To demonstrate the race condition you can open two connections to the same database, execute `BEGIN;` on connection 1 and fetch `SELECT pg_current_wal_lsn();`, we get an LSN that we want to use to start replication. On connection 2 now execute some DML statement (e.g. insert a row into a table). If we now fetch the current wal LSN on connection 1 again we will get a different LSN, which shows that the function does not return a consistent WAL LSN.

File: source/logrepl/iterator.go [hunk 1]
```diff
@@ -171,9 +217,31 @@ func (i *CDCIterator) attachSubscription(ctx context.Context, conn *pgx.Conn) er
 	return nil
 }
 
+// attachSnapshot checks if a snapshot should be taken and attaches
+// a snapshot iterator if it should. If Position is not set in the config, it
+// sets the current subscription's position to the snapshot's anchored LSN.
+func (i *Iterator) attachSnapshot(ctx context.Context, conn *pgx.Conn) error {
+	if i.config.SnapshotMode == "never" {
```
[the remark this answers is not in the document]
Comment (dylanlott) [reply source/logrepl/iterator.go:224]: I don't like magic constants either, I was planning on refactoring the config settings to be types and check for those but felt like it didn't belong in this PR. Noted for refactoring sooner, though.

File: source/logrepl/iterator.go [hunk 2]
```diff
@@ -80,10 +94,27 @@ func (i *CDCIterator) listen(ctx context.Context) {
...
-// block until either a record is returned from the subscription, the
-// subscription stops because of an error or the context gets canceled.
-func (i *CDCIterator) Next(ctx context.Context) (sdk.Record, error) {
+// Next checks if its snapshot is finished and returns a record from the
+// snapshot if it's not. If it is finished, it returns the next record retrieved
+// from the subscription. Once it's finished with the snapshot, it blocks until
+// either a record is returned from subscription, the subscription stops because
+// of an error, or the context is canceled.
+func (i *Iterator) Next(ctx context.Context) (sdk.Record, error) {
+	if !i.snap.Finished() {
```
[the remark this answers is not in the document]
Comment (dylanlott) [reply source/logrepl/iterator.go:103]: Good catch, will write a test for that.

File: source/logrepl/iterator.go [hunk 3]
```diff
@@ -138,7 +184,7 @@ func (i *CDCIterator) Teardown(ctx context.Context) error {
 
 // attachSubscription determines the starting LSN and key column of the source
 // table and prepares a subscription.
-func (i *CDCIterator) attachSubscription(ctx context.Context, conn *pgx.Conn) error {
+func (i *Iterator) attachSubscription(ctx context.Context, conn *pgx.Conn) error {
```
[the remark this answers is not in the document]
Comment (dylanlott) [reply source/logrepl/iterator.go:187]: I considered this exact case but I arrived at a different conclusion. My thought was that if a pipeline was stopped in the middle of operation, we would want to record the progress and thus the position, and that the manual intervention required would be necessary either way if they wanted to capture the snapshot. 

This is probably a good behavior to note in a connector development standards document somewhere.

[the remark this answers is not in the document]
Comment (lovromazgon) [reply source/logrepl/iterator.go:187]: Sure manual intervention will be needed, but let's define what "manual intervention"  means. Right now that would mean recreating the connector, otherwise the pipeline won't start, which is a bit much IMO.

I would say we should rather tell the user to check what records were written on the destination and clean it up (i.e. remove them if needed), because the connector will produce the same records again (or similar ones, in case the table content changed in the meantime). Conduit advertises "at least once delivery" so duplicated messages can already happen and destinations should ideally implement idempotent writes, so this behavior wouldn't violate any expectations.

TL;DR: I think the connector should behave the same as the S3 connector and, in case the position passed to `Open` is a snapshot position just restart the snapshot from scratch and print a warning that we can't restart an interrupted snapshot.
